streamlit/streamlit
# Fix nested fragment stale elements

## Describe your changes

Closes https://github.com/streamlit/streamlit/issues/9233
Closes https://github.com/streamlit/streamlit/issues/9267

This PR changes the way how we deprecate child nodes in fragment runs. 
Right now, we have the issue that child fragment elements can be stale even if the parent changed (related to issue https://github.com/streamlit/streamlit/issues/9233 but not unique to Dialogs). 

For the following videos, this app is used: 
```python
from uuid import uuid4

import streamlit as st

if "counter" not in st.session_state:
    st.session_state.counter = 0

@st.fragment
def outer_fragment():
    with st.container(border=True):
        st.session_state.counter += 1
        st.write(f"outer fragment: {uuid4()}")
        st.button("rerun outer fragment")
        inner_fragment()


@st.fragment
def inner_fragment():
    with st.container(border=True):
        st.write(f"inner fragment: {uuid4()}")
        st.button("rerun inner fragment")

        if 5 < st.session_state.counter < 10:
            st.write("FOO!")

st.write(f"outside all fragments: {uuid4()}")
st.button("rerun whole app")
outer_fragment()
```

Where the inner fragment writes an element when `session_state.count` is between 5 and 10. The count is incremented when the outer fragment is re-run. You can see that right now, even if the outer fragment is clicked more than 10 times, the `FOO!` does not disappear!

Before

https://github.com/user-attachments/assets/6ff73dcb-14bd-4a5f-ab8b-9078e6f94de4


After

https://github.com/user-attachments/assets/a319be75-0b56-4bba-90d5-d04083071e0b


---

The PR changes the behavior now as following:
- when the parent fragment was changed during a run, we consider all _nested_ fragment elements as stale as well unless they were also modified during the same run (the scriptRunId is correct).

---

An alternative approach to closing https://github.com/streamlit/streamlit/issues/923 would be to create a new function `cleanFragments` which is passed to the `Dialog` component and called in the `onClose` callback. But this would be a pure dialog-solution and wouldn't fix the general stale-element issue in fragments demonstrated in the videos.

## GitHub Issue Link (if applicable)

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
  - Add new frontend test to ensure that stale nested fragments are removed
- E2E Tests
  - Update the existing `st_fragments_nested` test.
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.
